Add batch author import script for STEX packages

timdmackey opened the pull request:
## Add `npm run add:author` script

- New script to batch importing all packages from one or more authors
- Queries STEX API by author name or ID
- Supports multiple authors in a single command
- Skips processing files which exist in the channel already, unless the `--update` option is specified

### Usage:

```bash
npm run add:author -- memo
npm run add:author -- memo "NAM Team"
npm run add:author -- 95442  # using author id
npm run add:author -- memo -u  # update existing files
```

### Additional Options:

- `-u, --update` - Update existing packages
- `-y, --yes` - Skip confirmation prompt

## Notes

- Exports run() function from manual-add.js for reuse in add-by-author.js

<img width="1384" height="976" alt="Screenshot 2025-11-01 at 6 47 38 PM" src="https://github.com/user-attachments/assets/beeb6c26-6251-4914-8696-65133083f9a6" />

timdmackey commented:
Accidentally jumped the gun in creating a PR for this. I still need to validate that the output files are being created correctly.

sebamarynissen commented on scripts/manual-add.js:
```diff
@@ -142,5 +143,8 @@ async function addFileToIndex(index, file) {
 
-const { argv } = yargs(hideBin(process.argv));
-await run(argv._, argv);
+// Only run when executed directly (not when imported as a module)
```
Clever trick. I always thought this was no longer possible with es modules. I used to do it like
```js
if (require.main === module) {}
```
in the CommonJS days and found this no longer being possible to be annoying. Thanks for this. 😄

timdmackey replied:
This was new for me too!

timdmackey commented:
> I never got to it because it would require a scraping approach and then handle pagination as well.

Needing to handle pagination here occurred to me as well, but I figured it's extremely unlikely any author will have 1,000 entries in the STEX.

> Small suggestion: I think it could be useful to add an option to the command to skip any files before a certain file id. Consider for example everything made by [IDS2](https://community.simtropolis.com/profile/70889-ids2/content/?type=downloads_file). A lot of their content is present in the channel, but since I've manually added all of that, they released additional content without a metadata.yaml file so the newest content is not yet available in the channel.

> It would be useful to be able to specify the last included file id to the script, and then skip anything before it. In a next step, you could then detect what the last file id is that's present in the channel and have it be taken into account automatically.

I assumed that this part of the `add()` function is accomplishing the same thing, but I haven't verified that. Am I understanding correctly that you're just wanting to exclude packages that are already in the channel?

https://github.com/sebamarynissen/simtropolis-channel/blob/23a78ad1255400c30622e6d45fb8a02ec107cd6d/scripts/manual-add.js#L25-L37

sebamarynissen commented:
It looks like that indeed, but actually it only filters out stuff that's in the default channel. Packages in the local channel are not filtered out (`pkg.local === true`), but adding the feature should just be a matter of modifying this filter function a bit.

timdmackey commented:
https://github.com/sebamarynissen/simtropolis-channel/blob/23a78ad1255400c30622e6d45fb8a02ec107cd6d/scripts/manual-add.js#L31
Yeah this logic seemed reversed based on my understanding of what should be happening. But it makes sense to want to handle updated files and not just ignore them.

Maybe I just add an argument which controls whether the script only looks at new files or looks at updated files too?

timdmackey commented:
I added an `--update` flag to both the scripts.

For `npm run add:author`, the default behaviour is to only import new files
For `npm run add`, since it's a more explicit command I set the default to update existing files

#### Default behavior:
- `npm run add <url...>` - adds or updates file
- `npm run add:author <author...>` - adds author files and ignores existing files

#### Override behavior:
- `npm run add <url...> --no-update` - adds file and ignores existing files
- `npm run add:author <author...> --update` - adds or updates author files

Note that `--update` has a shorthand `-u` for ease of use:
```
npm run add:author <author...> -u
```

I also added a confirmation dialog after the file list has been collected so you can check what will happen before proceeding:

<img width="1384" height="976" alt="Screenshot 2025-11-01 at 6 47 38 PM" src="https://github.com/user-attachments/assets/beeb6c26-6251-4914-8696-65133083f9a6" />

timdmackey commented:
@sebamarynissen This should be good to go now!

timdmackey commented:
> Small suggestion: I think it could be useful to add an option to the command to skip any files before a certain file id. Consider for example everything made by [IDS2](https://community.simtropolis.com/profile/70889-ids2/content/?type=downloads_file). A lot of their content is present in the channel, but since I've manually added all of that, they released additional content without a metadata.yaml file so the newest content is not yet available in the channel.

Actually I forgot about this, does my update flag solution solve the use case you were looking for?
